torch/_higher_order_ops/effects.py

```diff
@@ -123,6 +124,11 @@ def get_effect_key(op, args, kwargs) -> Optional[_EffectType]:
     return None
 
 
+def new_token_tensor() -> torch.Tensor:
+    # Use dtype bool to not affect Inductor dtype promotions
+    return torch.tensor([], dtype=torch.bool)
```

Comment:
hmm so if we leave the token's dtype as `float32`, inductor might try to do some type promotion and bitcast the token, causing it to end up in a runtime triton kernel?

I would have (naively) expected the token to be unused anywhere in the compute graph in inductor IR, and should (therefore) not show up in any kernels. Do you have a code snippet showing where leaving the token's type as `float32` breaks?

Reply:
Inductor lowering process before every lowering will `transform_args()`  https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L241
It will pick the highest dtype of all arguments, as a result of transform_args of with_effects(token, op, args) it will also count token.dtype and call promote for all args if their dtype is lower, e.g. args contain int32 dtype.

As a result it will create PointwiseOperation()

The dtype is picked here : https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L182

Technically we can improve this function to not count empty tensors.

Reply:
I'm ok with this as long as @eellison is. Although if we do then we should probably change all of the sites in this file that create tokens to use `torch.bool` too: https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/effects.py#L219